cmd/webhook-server/admission_controller.go

```diff
@@ -0,0 +1,155 @@
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, fmt.Errorf("invalid method %s, only POST requests are allowed", r.Method)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not read request body: %v", err)
+	}
+
+	if contentType := r.Header.Get("Content-Type"); contentType != jsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("unsupported content type %s, only %s is supported", contentType, jsonContentType)
+	}
+
+	// Step 2: Parse the AdmissionReview request.
+
+	var admissionReviewReq v1beta1.AdmissionReview
+
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not deserialize request: %v", err)
+	} else if admissionReviewReq.Request == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.New("malformed admission review: request is nil")
+	}
+
+	// Step 3: Construct the AdmissionReview response.
+
+	admissionReviewResponse := v1beta1.AdmissionReview{
+		Response: &v1beta1.AdmissionResponse{
+			UID: admissionReviewReq.Request.UID,
+		},
+	}
+
+	var patchOps []patchOperation
+	// Apply the admit() function only for non-Kubernetes namespaces. For objects in Kubernetes namespaces, return
+	// an empty set of patch operations.
+	if !isKubeNamespace(admissionReviewReq.Request.Namespace) {
+		patchOps, err = admit(admissionReviewReq.Request)
+	}
+
+	if err != nil {
+		// If the handler returned an error, incorporate the error message into the response and deny the object
+		// creation.
+		admissionReviewResponse.Response.Allowed = false
+		admissionReviewResponse.Response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+	} else {
+		// Otherwise, encode the patch operations to JSON and return a positive response.
+		_, err := json.Marshal(patchOps)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return nil, fmt.Errorf("could not marshal JSON patch: %v", err)
+		}
+		admissionReviewResponse.Response.Allowed = true
+
+	}
+
+	// Return the AdmissionReview with a response as JSON.
+	bytes, err := json.Marshal(&admissionReviewResponse)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling response: %v", err)
+	}
+	return bytes, nil
+}
+
+// serveAdmitFunc is a wrapper around doServeAdmitFunc that adds error handling and logging.
+func serveAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	log.Print("Handling webhook request ...")
```

Comment:
Try to use standard logger that we always use.

Reply:
Fixed